providers/implementations/keymgmt/rsa_kmgmt.c

```diff
@@ -0,0 +1,177 @@
+    BIGNUM *n = NULL, *e = NULL, *d = NULL;
+    STACK_OF(BIGNUM) *factors = NULL, *exps = NULL, *coeffs = NULL;
+
+    if (rsa == NULL)
+        return 0;
+
+    param_n = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_RSA_N);
+    param_e = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_RSA_E);
+    param_d = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_RSA_D);
+
+    if ((param_n != NULL && !OSSL_PARAM_get_BN(param_n, &n))
+        || (param_e != NULL && !OSSL_PARAM_get_BN(param_e, &e))
+        || (param_d != NULL && !OSSL_PARAM_get_BN(param_d, &d)))
+        goto err;
+
+    if (!RSA_set0_key(rsa, n, e, d))
+        goto err;
+    n = e = d = NULL;
+
+    if (!collect_numbers(factors = sk_BIGNUM_new_null(), params,
+                         OSSL_PKEY_PARAM_RSA_FACTOR)
+        || !collect_numbers(exps = sk_BIGNUM_new_null(), params,
+                            OSSL_PKEY_PARAM_RSA_EXPONENT)
+        || !collect_numbers(coeffs = sk_BIGNUM_new_null(), params,
+                            OSSL_PKEY_PARAM_RSA_COEFFICIENT))
+        goto err;
+
+    if (!RSA_set0_all_params(rsa, factors, exps, coeffs))
+        goto err;
+
+    sk_BIGNUM_free(factors);
+    sk_BIGNUM_free(exps);
+    sk_BIGNUM_free(coeffs);
+    return 1;
+
+ err:
+    BN_free(n);
+    BN_free(e);
+    BN_free(d);
+    sk_BIGNUM_pop_free(factors, BN_free);
+    sk_BIGNUM_pop_free(exps, BN_free);
+    sk_BIGNUM_pop_free(coeffs, BN_free);
+    return 0;
+}
+
+static int export_numbers(OSSL_PARAM params[], const char *key,
+                          STACK_OF(BIGNUM_const) *numbers)
+{
+    OSSL_PARAM *p = NULL;
+    int i, nnum;
+
+    if (numbers == NULL)
+        return 0;
+
+    nnum = sk_BIGNUM_const_num(numbers);
+
+    for (p = params, i = 0;
+         i < nnum && (p = OSSL_PARAM_locate(p, key)) != NULL;
+         p++, i++) {
+        if (!OSSL_PARAM_set_BN(p, sk_BIGNUM_const_value(numbers, i)))
+            return 0;
+    }
+
+    return i;
+}
+
+static int key_to_params(RSA *rsa, OSSL_PARAM params[])
+{
+    int ret = 0;
+    OSSL_PARAM *p;
+    const BIGNUM *rsa_d = NULL, *rsa_n = NULL, *rsa_e = NULL;
+    STACK_OF(BIGNUM_const) *factors = sk_BIGNUM_const_new_null();
+    STACK_OF(BIGNUM_const) *exps = sk_BIGNUM_const_new_null();
+    STACK_OF(BIGNUM_const) *coeffs = sk_BIGNUM_const_new_null();
+
+    if (rsa == NULL || factors == NULL || exps == NULL || coeffs == NULL)
+        goto err;
+
+    RSA_get0_key(rsa, &rsa_n, &rsa_e, &rsa_d);
+    RSA_get0_all_params(rsa, factors, exps, coeffs);
+
+    if ((p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_RSA_N)) != NULL
+        && !OSSL_PARAM_set_BN(p, rsa_n))
+        goto err;
+    if ((p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_RSA_E)) != NULL
+        && !OSSL_PARAM_set_BN(p, rsa_e))
+        goto err;
+    if ((p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_RSA_D)) != NULL
+        && !OSSL_PARAM_set_BN(p, rsa_d))
+        goto err;
+
+    if ((p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_RSA_PARAMS_NUM)) != NULL
```

Comment:
Is this actually necessary? We don't seem to use it in the equivalent `params_to_key` function

Reply:
It's gone now